### PR TITLE
Feat new usage stats

### PR DESCRIPTION
**Pull request by eldadfux**
## What does this PR do?

(Provide a description of what this PR does.)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work.)

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue link all related PR and issues here.)

### Have you added your change to the [Changelog](https://github.com/appwrite/appwrite/blob/master/CHANGES.md)?

(The CHANGES.md file tracks all the changes that make it to the `main` branch. Add your change to this file in the following format)
- One line description of your PR [#pr_number](Link to your PR)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

(Write your answer here.)

### CODE REVIEW

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:76]: where do we replace the `[DATABASE_ID]` with actual ID of the db?

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:244]: need to pass project and usage here

**Review by lohanidamodar — COMMENTED** [review]
Things are looking good, just one confusion regarding the namespace.

File: app/controllers/shared/api.php
Comment (eldadfux) [inline app/controllers/shared/api.php:92]: Please explain in comments what you do, every time you fetch stats.

File: app/worker.php
Comment (eldadfux) [inline app/worker.php:103]: Please avoid committing this change, it's much better to change it in your local env as this can be very risky if it gets merged.

File: composer.json
Comment (eldadfux) [inline composer.json:52]: Reminder to attach a tag.

File: src/Appwrite/Platform/Tasks/Usage.php
Comment (eldadfux) [inline src/Appwrite/Platform/Tasks/Usage.php:5]: This file should be removed.

File: src/Appwrite/Utopia/Response/Model/UsageProject.php
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/UsageProject.php:19]: We still need requests data.

File: docker-compose.yml
Comment (eldadfux) [inline docker-compose.yml:588]: We don't need the docker env vars, please validate we use only vars we need.

File: docker-compose.yml
Comment (eldadfux) [inline docker-compose.yml:96]: Please avoid committing this changes, they make tests break and are very hard to detect.

File: app/workers/usage.php
Comment (eldadfux) [inline app/workers/usage.php:52]: Indentation is wrong.

File: docker-compose.yml
Comment (shimonewman) [reply docker-compose.yml:562]: I left _APP_USAGE_STATS  for later as it appears in many places
and we do not have a final decision on whether to remove it.
I removed all other influx and telegraph occurrences.

File: composer.json
Comment (shimonewman) [reply composer.json:52]: Waiting database::increment branch to be mearged

File: bin/executor
Comment (lohanidamodar) [inline bin/executor:1]: i think this should no longer be here

**Review by lohanidamodar — CHANGES_REQUESTED** [review]
Went through everything in brief. The flow looks good to me. I'll review the implementation of listener again in more details.

File: bin/executor
Comment (shimonewman) [reply bin/executor:1]: I think it does, it runs php file via shell

File: app/executor.php
Comment (shimonewman) [reply app/executor.php:1]: Not sure i understand

Comment (lohanidamodar) [reply app/executor.php:1]: in the PR this file is shown as completely new file being added, probably because this was removed from `appwrite/appwrite` and moved to `open-runtimes` please consult @Meldiron

File: bin/executor
Comment (lohanidamodar) [reply bin/executor:1]: I mean the I think the whole `executor` was moved to open-runtimes, @Meldiron knows better

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:52]: Let's also make sure we have proper error handling inside this block so that if this fails, we log the issue and note break the API

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:169]: instead of doing the decrement here, why not just add this also to the queue?

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:177]: Should we also have a `storage` metric that will account for `files.storage` + `deployments.storage` + ... later (build.storage, statichosting.stroage) and such? or do we calculate that during runtime in billing?

Comment (lohanidamodar) [reply app/controllers/shared/api.php:176]: @eldadfux suggested, let's keep it separate it's fine to do this on the billing itself

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:193]: Namespace should n ot be `appwrite-worker` as this is not worker, it's API

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:189]: what if error code is not `500` ? again that will fail the API?

File: app/controllers/shared/api.php
Comment (lohanidamodar) [inline app/controllers/shared/api.php:201]: can we also make sure, we add event, so we know which event failed as well

File: app/controllers/shared/api.php
Comment (shimonewman) [reply app/controllers/shared/api.php:189]: ok

File: app/controllers/shared/api.php
Comment (shimonewman) [reply app/controllers/shared/api.php:201]: if you know the collection then you know what metric failed, am i wrong?

File: .env
Comment (christyjacob4) [inline .env:55]: can we revert this back ?

File: Dockerfile
Comment (christyjacob4) [inline Dockerfile:308]: We don't need this anymore . The executor has been moved to open runtimes

File: tests/e2e/General/AbuseTest.php
Comment (christyjacob4) [inline tests/e2e/General/AbuseTest.php:20]: any reason we changed this ?

File: tests/benchmarks/ws.js
Comment (christyjacob4) [inline tests/benchmarks/ws.js:2]: are we not using these files anywhere?

File: tests/benchmarks/http.js
Comment (christyjacob4) [inline tests/benchmarks/http.js:20]: are we not using this file anywhere?

File: bin/executor
Comment (christyjacob4) [inline bin/executor:3]: This file is not required since we moved the executor to open-runtimes.

File: phpbench.json
Comment (christyjacob4) [inline phpbench.json:2]: You can delete this

File: app/executor.php
Comment (shimonewman) [reply app/executor.php]: removed

File: bin/executor
Comment (shimonewman) [reply bin/executor]: removed

File: app/executor.php
Comment (shimonewman) [reply app/executor.php]: done

File: .env
Comment (shimonewman) [reply .env:55]: Done

File: tests/benchmarks/ws.js
Comment (shimonewman) [reply tests/benchmarks/ws.js]: we don't , removed

File: bin/executor
Comment (shimonewman) [reply bin/executor]: removed

File: phpbench.json
Comment (shimonewman) [reply phpbench.json]: deleted

File: tests/benchmarks/http.js
Comment (shimonewman) [reply tests/benchmarks/http.js]: reverted

File: Dockerfile
Comment (shimonewman) [reply Dockerfile]: removed

File: tests/e2e/General/AbuseTest.php
Comment (shimonewman) [reply tests/e2e/General/AbuseTest.php]: reverted

File: app/config/usage.php
Comment (christyjacob4) [inline app/config/usage.php:7]: What is the factor used for ?

File: app/controllers/api/account.php
Comment (christyjacob4) [inline app/controllers/api/account.php:148]: If we removed these labels, how is the usage being calculated?

File: app/controllers/api/databases.php
Comment (christyjacob4) [inline app/controllers/api/databases.php:854]: Any reason for this change ?

File: app/controllers/api/databases.php
Comment (christyjacob4) [inline app/controllers/api/databases.php:2539]: This end point was originally present in Line 2450. Can you please move it back there so the diff is easier to read ?

File: app/views/install/compose.phtml
Comment (christyjacob4) [inline app/views/install/compose.phtml:433]: Can we fix this diff? 

`appwrite-maintenance` was already part of the compose file.. It need not be modified to `appwrite-worker-usage`

`appwrite-worker-usage` can be added as a new item for a cleaner diff

File: app/workers/builds.php
Comment (christyjacob4) [inline app/workers/builds.php:262]: We need to define these metrics as constants somewhere. Else they will get very hard to manage

BUILDS_COUNT = 'builds';
BUILDS_STORAGE = 'builds.storage';
BUILDS_COMPUTE = 'builds.compute';


FUNCTION_BUILDS_COUNT = '{functionInternalId}.builds';

and so on..

Comment (christyjacob4) [reply app/workers/builds.php:262]: I also see that we're missing the `buildStatus` metric.

File: composer.json
Comment (christyjacob4) [inline composer.json:66]:
```suggestion
        "utopia-php/storage": "0.13.*",
```

File: docker-compose.yml
Comment (christyjacob4) [inline docker-compose.yml:56]: This line is required

File: docker-compose.yml
Comment (christyjacob4) [inline docker-compose.yml:550]: can you revert this back to `appwrite-maintenance`

File: docker-compose.yml
Comment (christyjacob4) [inline docker-compose.yml:587]: You can modify this section from `appwrite-usage` to `appwrite-worker-usage`

File: src/Appwrite/Platform/Services/Tasks.php
Comment (christyjacob4) [inline src/Appwrite/Platform/Services/Tasks.php:27]: You can delete the import as well

File: src/Appwrite/Platform/Tasks/Doctor.php
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/Doctor.php:209]: If we removed the check for influx and telegraf, we still need to add checks for the connection to the usage queue

File: tests/e2e/Services/Databases/DatabasesConsoleClientTest.php
Comment (christyjacob4) [inline tests/e2e/Services/Databases/DatabasesConsoleClientTest.php:90]: what about all the other metrics ? 👀

File: app/config/collections.php
Comment (christyjacob4) [inline app/config/collections.php:2785]: I believe all these  attributes need to marked `required`

File: app/controllers/api/functions.php
Comment (christyjacob4) [inline app/controllers/api/functions.php:1366]: Can you please keep this endpoint back on Line 211 ? Else this diff is hard to interpret .

File: app/controllers/api/functions.php
Comment (christyjacob4) [inline app/controllers/api/functions.php:1450]: Can you please keep this endpoint back on Line 321 ? Else this diff is hard to interpret .

File: app/config/usage.php
Comment (shimonewman) [reply app/config/usage.php:7]: To calculate the leaps on the API endpoints (for instance the storage usage endpoint)

image

File: app/controllers/api/project.php
Comment (christyjacob4) [inline app/controllers/api/project.php:40]: We need to use constants here

File: app/controllers/shared/api/auth.php
Comment (christyjacob4) [inline app/controllers/shared/api/auth.php:10]: Why did we move this to a separate file ?

File: app/controllers/shared/api.php
Comment (christyjacob4) [inline app/controllers/shared/api.php:284]: Why don't we set these listeners in the `App::setResource('dbForProject')`  ?

File: src/Appwrite/Utopia/Response/Model/UsageFunction.php
Comment (christyjacob4) [inline src/Appwrite/Utopia/Response/Model/UsageFunction.php:19]: We use `documentsCount` and `collectionsCount` in the database usage and only `deployments`, `builds` etc in the functions usage. We need to maintain consistency

File: src/Appwrite/Utopia/Response/Model/UsageFunctions.php
Comment (christyjacob4) [inline src/Appwrite/Utopia/Response/Model/UsageFunctions.php:19]: Same as my comment above. We need to maintain consistency in the way we name the metrics. we use the `Count` suffix in some places and don't use it in some..

File: src/Appwrite/Utopia/Response/Model/UsageBuckets.php
Comment (shimonewman) [reply src/Appwrite/Utopia/Response/Model/UsageBuckets.php:60]: Eldad deprecated these metrics

Comment (christyjacob4) [reply src/Appwrite/Utopia/Response/Model/UsageBuckets.php:60]: So how do we get this information if we want to? How will the usage UI change once we merge this? We will not have these graphs anymore 🤔

File: app/controllers/api/functions.php
Comment (shimonewman) [reply app/controllers/api/functions.php:1450]: ok

File: app/controllers/api/functions.php
Comment (shimonewman) [reply app/controllers/api/functions.php:1366]: ok

File: tests/e2e/Services/Databases/DatabasesConsoleClientTest.php
Comment (shimonewman) [reply tests/e2e/Services/Databases/DatabasesConsoleClientTest.php:90]: They where deprecated

File: docs/specs/overview.drawio.svg
Comment (christyjacob4) [inline docs/specs/overview.drawio.svg:592]: Can we remove the telegraf and statsd containers completely?

File: src/Appwrite/Utopia/Response/Model/UsageBuckets.php
Comment (shimonewman) [reply src/Appwrite/Utopia/Response/Model/UsageBuckets.php:60]: They will need to make UI changes

File: app/controllers/api/account.php
Comment (christyjacob4) [reply app/controllers/api/account.php:148]: Okay 👍

File: app/controllers/shared/api/auth.php
Comment (eldadfux) [reply app/controllers/shared/api/auth.php:10]: Yeh, I wanted to try and minimize each hook's responsibility

Comment (christyjacob4) [reply app/controllers/shared/api/auth.php:10]: Okay sounds good. @shimonewman let's keep it that way 👍

File: app/controllers/api/databases.php
Comment (shimonewman) [reply app/controllers/api/databases.php]: reverted

File: app/controllers/shared/api.php
Comment (christyjacob4) [inline app/controllers/shared/api.php:172]: We already have a logger method in this file with all the default params set. I don't think we need to reinitialise it

File: docker-compose.yml
Comment (christyjacob4) [inline docker-compose.yml:828]:
```suggestion
```

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**
